File: src/cogs/manager/manager.py
import asyncio
import http.client

import interactions
from interactions import Extension, SlashContext

from lib.util import subcommand, IS_ADMIN


class Manager(Extension):
    '''Commands for managing the discord server'''

    @subcommand(channel={'description': 'The channel to send the message in', 'channel_types': [interactions.ChannelType.GUILD_TEXT]})
    @IS_ADMIN
    async def say(self, ctx: SlashContext, channel: interactions.GuildText) -> None:
        '''
        Says your previous message in the channel you specify
        '''
        await ctx.defer()
        try:
            messages = await ctx.channel.history(limit=100).flatten()
            message = [m for m in messages if m.author.id == ctx.author.id][0].content
        except IndexError:
            await ctx.send(":x: Send the message in the current channel before calling /manager edit.")
            return

        confirm_message = await ctx.send(f'Should I send this in {channel.mention}? (60s expiry):\n{message}\n')

        await confirm_message.add_reaction('✅')

        def check(event: interactions.events.MessageReactionAdd):
            return event.message.id == confirm_message.id and \
                event.author == ctx.author and event.emoji.name == '✅'
        try:
            await self.bot.wait_for(interactions.events.MessageReactionAdd, checks=check, timeout=60.0)
        except asyncio.TimeoutError:
            await ctx.send(':x: Timed out, did not send.')
            return

        await channel.send(message)
        await ctx.send(f':white_check_mark: sent message in {channel.mention}')

    @subcommand(message_channel={'description': 'The channel of the original message', 'channel_types': [interactions.ChannelType.GUILD_TEXT]}, message_id={'description': 'The ID of the message to edit'})
    @IS_ADMIN
    async def edit(self, ctx: SlashContext, message_channel: interactions.GuildText, message_id: str) -> None:
        '''
        Edits a message said by the bot, must specify the messageID 
        '''
        await ctx.defer()
        try:
            bot_message = await message_channel.fetch_message(message_id)
        except (http.client.HTTPException, ValueError):
            await ctx.send(":x: Invalid message_id")
            return
        if (bot_message is None):
            await ctx.send(":x: Unable to find message")
            return
        if (bot_message.author != self.bot.user):
            await ctx.send(":x: Message must be from the bot")
            return

        try:
            messages = await ctx.channel.history(limit=100).flatten()
            message = [m for m in messages if m.author.id == ctx.author.id][0].content
        except IndexError:
            await ctx.send(":x: Send the message in the current channel before calling /manager edit.")
            return

        confirm_message = await ctx.send(f'Should I edit the message you specified to say this? (60s expiry):\n{message}\n')

        await confirm_message.add_reaction('✅')

        def check(event: interactions.events.MessageReactionAdd):
            return event.message.id == confirm_message.id and \
                event.author == ctx.author and event.emoji.name == '✅'
        try:
            await self.bot.wait_for(interactions.events.MessageReactionAdd, checks=check, timeout=60.0)
        except asyncio.TimeoutError:
            await ctx.send(':x: Timed out, did not edit message.')
            return

        await bot_message.edit(content=message)
        await ctx.send(f':white_check_mark: edited message.')

    @subcommand(
        role_name={'description': 'The name of the role to assign'},
        usernames={'description': 'A comma-separated list of usernames (e.g. "user1, user2")'}
    )
    @IS_ADMIN
    async def assign_roles(self, ctx: SlashContext, role_name: str, usernames: str) -> None:
        '''
        Assigns a role to a list of users by username.
        '''
        await ctx.defer()

        user_list = [u.strip() for u in usernames.replace(", ", ",").split(",") if u.strip()]
        if not user_list:
            await ctx.send(":x: Please provide a comma-separated list of usernames.")
            return

        guild = ctx.guild
        if guild is None:
            await ctx.send(":x: Could not determine the guild context.")
            return
        
        role = next((r for r in guild.roles if r.name == role_name), None)
        if not role:
            await ctx.send(f":x: Role '{role_name}' not found.")
            return

        assigned = []
        not_found = []
        for username in user_list:
            member = next((m for m in guild.members if m.user.username == username), None)
            if member:
                try:
                    await member.add_role(role)
                    assigned.append(username)
                except Exception:
                    not_found.append(username)
            else:
                not_found.append(username)

        msg = ""
        if assigned:
            msg += f":white_check_mark: Assigned role '{role_name}' to: {', '.join(assigned)}\n"
        if not_found:
            msg += f":x: Could not find or assign role to: {', '.join(not_found)}"
        await ctx.send(msg or ":x: No users processed.")

File: server/src/commands/manager.ts
import {
  ApplicationCommandOptionType,
  ApplicationCommandType,
  ButtonStyle,
  ChannelType,
  ComponentType,
  PermissionFlagsBits,
} from "discord-api-types/v10";
import {
  addGuildMemberRole,
  editOriginalInteractionResponseWithFile,
  listChannelMessages,
  listGuildRoles,
  searchGuildMembers,
} from "../discord/client.js";
import { message, modal } from "../discord/responses.js";
import { deferredTask, finish } from "./helpers.js";
import {
  channelId,
  guildId,
  stringOption,
  subcommand,
  userId,
} from "./options.js";
import type { CommandDefinition } from "./types.js";

const ACTION_TTL_MS = 60_000;
const MODAL_TTL_MS = 15 * 60_000;

export const managerCommand: CommandDefinition = {
  command: {
    name: "manager",
    type: ApplicationCommandType.ChatInput,
    description: "No Description Set",
    default_member_permissions: String(PermissionFlagsBits.Administrator),
    options: [
      {
        type: ApplicationCommandOptionType.Subcommand,
        name: "say",
        description: "Says your previous message in the channel you specify",
        options: [
          {
            type: ApplicationCommandOptionType.Channel,
            name: "channel",
            description: "The channel to send the message in",
            required: true,
            channel_types: [ChannelType.GuildText],
          },
        ],
      },
      {
        type: ApplicationCommandOptionType.Subcommand,
        name: "edit",
        description: "Edits a message said by the bot",
        options: [
          {
            type: ApplicationCommandOptionType.Channel,
            name: "message_channel",
            description: "The channel of the original message",
            required: true,
            channel_types: [ChannelType.GuildText],
          },
          {
            type: ApplicationCommandOptionType.String,
            name: "message_id",
            description: "The ID of the message to edit",
            required: true,
          },
        ],
      },
      {
        type: ApplicationCommandOptionType.Subcommand,
        name: "assign_roles",
        description: "Assigns a role to a list of users by username.",
        options: [
          {
            type: ApplicationCommandOptionType.String,
            name: "role_name",
            description: "The name of the role to assign",
            required: true,
          },
          {
            type: ApplicationCommandOptionType.String,
            name: "usernames",
            description:
              'A comma-separated list of usernames (e.g. "user1, user2")',
            required: true,
          },
        ],
      },
    ],
  },
  async handle(interaction, runtime) {
    const targetGuildId = guildId(interaction);
    const invocationChannelId = channelId(interaction);
    if (!targetGuildId || !invocationChannelId) {
      return message(":x: Could not determine the guild context.");
    }

    const action = subcommand(interaction)?.name;
    if (action === "say") {
      return deferredTask(interaction, runtime, true, async () => {
        const authorId = userId(interaction);
        const source = (
          await listChannelMessages(runtime.env, invocationChannelId)
        ).find((candidate) => candidate.author.id === authorId);
        if (!source?.content) {
          await finish(
            interaction,
            runtime,
            ":x: Send the message in the current channel before calling /manager say.",
          );
          return;
        }

        const id = crypto.randomUUID();
        const now = Date.now();
        const targetChannelId = stringOption(interaction, "channel");
        await runtime.storage.createInteractionState({
          id,
          kind: "manager_say",
          ownerUserId: authorId,
          guildId: targetGuildId,
          channelId: invocationChannelId,
          payload: { targetChannelId, content: source.content },
          createdAt: now,
          expiresAt: now + ACTION_TTL_MS,
        });
        await finish(interaction, runtime, {
          content: `Should I send this in <#${targetChannelId}>? (60s expiry)`,
          embeds: [{ description: source.content }],
          components: [
            {
              type: ComponentType.ActionRow,
              components: [
                {
                  type: ComponentType.Button,
                  style: ButtonStyle.Success,
                  label: "Confirm",
                  custom_id: `manager-say:confirm:${id}`,
                },
                {
                  type: ComponentType.Button,
                  style: ButtonStyle.Secondary,
                  label: "Cancel",
                  custom_id: `manager-say:cancel:${id}`,
                },
              ],
            },
          ],
        });
      });
    }

    if (action === "edit") {
      const id = interaction.id;
      const now = Date.now();
      await runtime.storage.createInteractionState({
        id,
        kind: "manager_edit",
        ownerUserId: userId(interaction),
        guildId: targetGuildId,
        channelId: invocationChannelId,
        payload: {
          targetChannelId: stringOption(interaction, "message_channel"),
          targetMessageId: stringOption(interaction, "message_id"),
        },
        createdAt: now,
        expiresAt: now + MODAL_TTL_MS,
      });
      return modal(
        `manager-edit:${id}`,
        "Edit bot message",
        "Replacement message",
      );
    }

    const roleName = stringOption(interaction, "role_name");
    const usernames = stringOption(interaction, "usernames")
      .replaceAll(", ", ",")
      .split(",")
      .map((username) => username.trim())
      .filter(Boolean);
    if (usernames.length === 0) {
      return message(":x: Please provide a comma-separated list of usernames.");
    }

    return deferredTask(interaction, runtime, false, async () => {
      const role = (await listGuildRoles(runtime.env, targetGuildId)).find(
        (candidate) => candidate.name.toLowerCase() === roleName.toLowerCase(),
      );
      if (!role) {
        await finish(interaction, runtime, `:x: Role '${roleName}' not found.`);
        return;
      }

      const assigned: string[] = [];
      const notFound: string[] = [];
      for (const username of usernames) {
        const candidates = await searchGuildMembers(
          runtime.env,
          targetGuildId,
          username,
        );
        const member = candidates.find(
          (candidate) =>
            candidate.user?.username.toLowerCase() === username.toLowerCase(),
        );
        if (!member?.user) {
          notFound.push(username);
          continue;
        }
        try {
          await addGuildMemberRole(
            runtime.env,
            targetGuildId,
            member.user.id,
            role.id,
          );
          assigned.push(username);
        } catch {
          notFound.push(username);
        }
      }

      const details = [
        assigned.length
          ? `Assigned role '${role.name}' to: ${assigned.join(", ")}`
          : "No roles assigned.",
        notFound.length
          ? `Could not find or assign role to: ${notFound.join(", ")}`
          : "No assignment failures.",
      ].join("\n");
      const summary = `:white_check_mark: Assigned: ${assigned.length}\n:x: Not assigned: ${notFound.length}`;
      if (details.length > 1_900) {
        await editOriginalInteractionResponseWithFile(
          runtime.env,
          interaction.token,
          { content: summary },
          "role-assignment-report.txt",
          details,
        );
      } else {
        await finish(interaction, runtime, `${summary}\n${details}`);
      }
    });
  },
};
